File: 10204 - Stack 'em up/stackemup.cpp
/* @JUDGE_ID:   1318761   10204     CPP     */

/* @BEGIN_OF_SOURCE_CODE */

#include <iostream>
#include <string>
#include <vector>

struct Shuffle {
    char order[52];

    Shuffle() {
        for (auto i = 0; i < 52; i++) {
            order[i] = 0;
        }
        return;
    }

    ~Shuffle() {
        return;
    }

};

std::string cardName(char value);

int main() {
    std::size_t space = 0; // Helper var used for parsing strings
    std::string input = "";
    char alpha_deck[52]; // main working deck
    char swap_deck[52]; // temporary deck used to hold swapped order
    bool shuffling = 0;
    std::vector<Shuffle> shuffles;
    int test_cases = 0; // number of test cases
    int shuffle_count = 0; // number of shuffles in the test case

    // get first line of input, test case count
    std::getline(std::cin, input);

    test_cases = std::stoi(input);

    for (auto i = 0; i < test_cases; i++) {
        // re-initialize deck for every test case
        for (auto i = 0; i < 52 ; i++) {
            alpha_deck[i] = i;
        }

        // get first line of test case, number of shuffles

        do {
            std::getline(std::cin, input);
        } while (input == "");

        shuffle_count = std::stoi(input);

        // read in shuffles.
        for (auto j = 0; j < shuffle_count; j++) {
            std::getline(std::cin, input);
            shuffles.push_back(*(new Shuffle()));
            for (auto k = 0; k < 52; k++) {
                if (input == "") {
                    std::getline(std::cin, input);
                } else {
                    // do_nothing
                }
                shuffles[j].order[k] = std::stoi(input, &space) - 1;
                input = input.substr(space);
            }
        }

        // read out order
        shuffling = true;
        int set = 0;

        std::getline(std::cin, input);
        set = std::stoi(input) - 1;
        do {
            for (auto j = 0; j < 52; j++) {
                swap_deck[j] = alpha_deck[static_cast<int>(shuffles[set].order[j])];
            }
            for (auto j = 0; j < 52; j++) {
                alpha_deck[j] = swap_deck[j];
            }
            std::getline(std::cin, input);
            if (input == "") {
                shuffling = false;
            } else {
                set = std::stoi(input) - 1;
            }
            
        } while (shuffling);
        //output final order
        for (auto j: alpha_deck) {
            std::cout << cardName(j) << '\n';
        }
        if (i < (test_cases - 1)) {
            std::cout << '\n' << std::flush;
        }
        shuffles.clear();
    }
    return 0;
}

std::string cardName(char value) {
    std::string result = "";
    int card = value % 13;

    switch (card) {
        case 0:
            result = "2";
            break;
        case 1:
            result = "3";
            break;
        case 2:
            result = "4";
            break;
        case 3:
            result = "5";
            break;
        case 4:
            result = "6";
            break;
        case 5:
            result = "7";
            break;
        case 6:
            result = "8";
            break;
        case 7:
            result = "9";
            break;
        case 8:
            result = "10";
            break;
        case 9:
            result = "Jack";
            break;
        case 10:
            result = "Queen";
            break;
        case 11:
            result = "King";
            break;
        case 12:
            result = "Ace";
            break;
        default:
            result = "N/A";
            break;
    }

    result += " of ";

    if ( value <= 12) {
        result += "Clubs";
    } else if (value > 12 && value <= 25) {
        result += "Diamonds";
    } else if (value > 25 && value <= 38) {
        result += "Hearts";
    } else if (value > 38) {
        result += "Spades";
    }

    return result;
}

/* @END_OF_SOURCE_CODE */

File: 10205 - Stack 'em up/stackemup.pas
program stackemup(input, output);
uses
    sysUtils;
type
    deckArray = array[0..51] of char;
    // Could technically have made this an object, or a class, but it did 
    //   not want to cooperate- I had issues initializing an array of objects.
    //   Thankfully, records behave better, even if I need to explicitly initia-
    //   ize them in main. It does not seem to hurt performance any- The same
    //   work would be done either way.
    Shuffle = record
        order : deckArray;
    end; 
    ShuffleArr = array of Shuffle;

var
    space : sizeInt;
    line : ansistring; // Holds a copy of our current input line.
    alpha_deck : deckArray;
    swap_deck : deckArray;
    shuffling : boolean;
    shuffles : ShuffleArr;
    test_cases : longInt; // Number of test cases.
    shuffle_count : longInt; // Number of shuffles in the test case.
    set_shuffle : ShortInt;
    i, j, l : longInt; // Counter Variables

    function delim (var str : ansistring) : sizeInt;
    var
        result, counter, size : sizeInt;
    begin 
        size := Length(str);
        counter := 1;
        result := size + 1;
        while counter < size do
        begin
            if Copy(str,counter,1) = ' ' then
            begin
                result := counter;
                counter := size;
            end;
        counter := counter + 1;
        end;
        delim := result;
    end;

    function cardName (value : char) : ansistring;
    var
        result : ansistring;
        card : longInt;
    begin
        card := longInt(value) mod 13;

        case (card) of
            0 : result := '2';
            1 : result := '3';
            2 : result := '4';
            3 : result := '5';
            4 : result := '6';
            5 : result := '7';
            6 : result := '8';
            7 : result := '9';
            8 : result := '10';
            9 : result := 'Jack';
            10 : result := 'Queen';
            11 : result := 'King';
            12 : result := 'Ace';
        else
            result := 'Oh, dear!';
        end;

        result := result + ' of ';

        if value <= char(12) then
            result := result + 'Clubs'
        else if (value > char(12)) and (value <= char(25)) then
            result := result + 'Diamonds'
        else if (value > char(25)) and (value <= char(38)) then
            result := result + 'Hearts'
        else if value > char(38) then
            result := result + 'Spades';

        cardName := result;
    end;
begin
    // Get first line of input, test case count.
    readln(line);
    test_cases := StrToInt(line);
    // Main for loop- Originally from i = 0 to (i < test_cases)
    for l := 0 to test_cases - 1 do
    begin
        // re-initialize deck for every test case
        for i := 0 to 51 do 
        begin
            alpha_deck[i] := char(i);
            swap_deck[i] := char(i);
        end;

        // Get first line of test case, number of shuffles.
        repeat readln(line) until not (line = '');

        shuffle_count := StrToInt(line);
        setlength(shuffles, shuffle_count);
        for i := Low(shuffles) to High(shuffles) do
            for j := Low(shuffles[i].order) to High(shuffles[i].order) do
                shuffles[i].order[j] := char(0);

        // read in shuffles

        for i := 0 to shuffle_count - 1 do
        begin
            readln(line);
            for j := 0 to 51 do 
            begin
                if (line = '') then readln(line);
                space := delim(line);
                shuffles[i].order[j] := char(StrToInt(Copy(line, 0, space - 1)) - 1);
                Delete(line, 1, space);
            end;
        end;

        // read out order
        shuffling := true;
        set_shuffle := 0;

        readln(line);
        set_shuffle := StrToInt(line) - 1;
        repeat
            for i := 0 to 51 do
                swap_deck[i] := alpha_deck[ShortInt(shuffles[set_shuffle].order[i])];
            
            for i := 0 to 51 do
                alpha_deck[i] := swap_deck[i];
            
            readln(line);
            if line = '' then shuffling := false 
                else set_shuffle := StrToInt(line) - 1;

        until not shuffling;

        // output final order
        for i := Low(alpha_deck) to High(alpha_deck) do
            writeln(cardName(alpha_deck[i]));
        
        if l < (test_cases) - 1 then
            writeln('');
        
    end;
end.